### PR TITLE
Update authenticator CR with status and events

### PR DESCRIPTION
**Pull request by pperiyasamy**
This enhances authenticator crd with status field containing information about every listening interfaces for hostapd process and its associated authenticated stations.
It also provides facility for recording events for monitor module and records events for authenticator crd in the scenarios like supplicant auth failures, interface operational state changes

### CODE REVIEW

**Comment by pperiyasamy** [comment]
The commit 33b1f73c7fd0fde26d678e07fc6c30720f84d3e1 get authenticator CR populated with status.

```
apiVersion: eapol.eapol.openshift.io/v1
kind: Authenticator
metadata:
  creationTimestamp: "2023-04-20T15:28:48Z"
  generation: 1
  name: authenticator-sample
  namespace: default
  resourceVersion: "1790325"
  uid: 8dcd0893-9c8f-4de8-9fb2-20c708eb214d
spec:
  authentication:
    local:
      authPort: 1812
      caCertSecret:
        name: localauth-2hcbd4mg4m
      privateKeyPassphrase: Hsecret
      privateKeySecret:
        name: localauth-2hcbd4mg4m
      radiusClientFileSecret:
        name: localauth-2hcbd4mg4m
      serverCertSecret:
        name: localauth-2hcbd4mg4m
      userFileSecret:
        name: localauth-2hcbd4mg4m
  enabled: true
  interfaces:
  - enp0s10
  serviceAccount: authenticator
status:
  interfaces:
  - authenticatedClients:
    - 08:00:27:6f:cd:84
    name: enp0s10
    status: ENABLED

```

**Comment by pperiyasamy** [comment]
The commit 51177b46036ab4ec4ecda80ea80c274158ba2a28 records event for authenticator crd as like below.

```
Spec:
  Authentication:
    Local:
      Auth Port:  1812
      Ca Cert Secret:
        Name:                  localauth-2hcbd4mg4m
      Private Key Passphrase:  Hsecret
      Private Key Secret:
        Name:  localauth-2hcbd4mg4m
      Radius Client File Secret:
        Name:  localauth-2hcbd4mg4m
      Server Cert Secret:
        Name:  localauth-2hcbd4mg4m
      User File Secret:
        Name:  localauth-2hcbd4mg4m
  Enabled:     true
  Interfaces:
    enp0s10
  Service Account:  authenticator
Status:
  Interfaces:
    Authenticated Clients:
    Name:    enp0s10
    Status:  ENABLED
Events:
  Type     Reason   Age   From           Message
  ----     ------   ----  ----           -------
  Warning  enp0s10  52s   authenticator  authentication failure for supplicant 08:00:27:6f:cd:84
  Warning  enp0s10  7s    authenticator  interface is down
  Normal   enp0s10  3s    authenticator  interface is up
```

File: api/v1/authenticator_types.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ type AuthenticatorSpec struct {
 	// Authentication configures back-end authentication for this authenticator
 	Authentication Auth `json:"authentication"`
 
+	// ServiceAccount sa to used by authenticator daemonset to update status
+	ServiceAccount string `json:"serviceAccount"`
```
Comment (lack) [inline api/v1/authenticator_types.go:43]: Does this need to be configurable by the end-user and put in every CR?  Can this just be handled internally by the controller?

Comment (lack) [inline api/v1/authenticator_types.go]: Does this need to be configurable by the end-user and put in every CR? Can this just be handled internally by the controller?

File: internal/k8s/k8s.go
Comment (lack) [inline internal/k8s/k8s.go:1]: This looks like boilerplate code - Is it copied in from somewhere else?

File: pkg/configgen/configgen.go [hunk 2]
```diff
@@ -350,3 +366,11 @@ func (g *ConfigGenerator) parsePorts() (string, string) {
 	}
 	return tcp.String(), udp.String()
 }
+
+// NamespacedName returns the general purpose string representation
+func NamespacedName(authenticator *eapolv1.Authenticator) string {
```
Comment (lack) [inline pkg/configgen/configgen.go:371]: This feels like there's already some implementation somewhere we could reuse, maybe?

File: api/v1/authenticator_types.go [hunk 3]
```diff
@@ -150,6 +153,20 @@ type Ports struct {
...
+
+	// Interfaces is the list of interface status
+	// +optional
+	Interfaces []*Interface `json:"interfaces,omitempty"`
+}
+
+type Interface struct {
+	// Name is the name of the interface
+	Name string `json:"name"`
+	// State is the state of the interface
```
Comment (lack) [inline api/v1/authenticator_types.go]: If these state values are from somewhere (from hostapd itself?) we should mention what the expected values are.

File: monitor/main.go [hunk 4]
```diff
@@ -88,8 +106,8 @@ func main() {
 	var monitors []*hostap.InterfaceMonitor
 	for _, intf := range ifaces {
 		level.Info(logger).Log("op", "startup", "monitor start for interface", intf)
-		intfMonitor := &hostap.InterfaceMonitor{Logger: logger, IfName: intf,
-			IfEventHandler: ifEventHandler}
+		intfMonitor := &hostap.InterfaceMonitor{Logger: logger, Client: k8Client, Recorder: eventRecorder,
+			AuthNsName: authObjKey, IfName: intf, IfEventHandler: ifEventHandler}
```
Comment (lack) [inline monitor/main.go]: Maybe it's time to split this into a multi-line initializer?

File: pkg/hostap/monitor.go [hunk 5]
```diff
@@ -140,17 +157,30 @@ func (m *InterfaceMonitor) handleHostapdEvent(eventStr string) error {
 	eventStrSlice := strings.Split(eventStr, " ")
 	if len(eventStrSlice) < 2 {
 		if isSolicitedEvent(eventStr) {
+			if strings.Contains(eventStr, statusReply) {
+				m.ifEAPState = getState(eventStr)
```
Comment (lack) [inline pkg/hostap/monitor.go]: This feels like a bit of an API bleed; we're exposing part of the raw event from hostapd to the end-user; Would it be safer to come up with our own state enum instead?

File: pkg/hostap/monitor.go [hunk 6]
```diff
@@ -140,17 +157,30 @@ func (m *InterfaceMonitor) handleHostapdEvent(eventStr string) error {
...
+	}()
 	eventKeyStr := eventStrSlice[0][strings.Index(eventStrSlice[0], ">")+1:]
 	switch eventKeyStr {
 	case staConnectedEvent, eapSuccessEvent:
 		return m.handleAuthenticateEvent(eventStrSlice[1])
-	case staDisconnectedEvent, eapFailureEvent:
+	case staDisconnectedEvent:
 		return m.handleDeAuthenticateEvent(eventStrSlice[1])
+	case eapFailureEvent:
+		m.logEvent(kapi.EventTypeWarning, "authentication failure for supplicant %s", eventStrSlice[1])
```
Comment (lack) [inline pkg/hostap/monitor.go:187]: I see we're logging a k8s Event for auth failure, interface up, interface down, but I don't see one for authentication or deauthentication.

File: pkg/configgen/configgen.go [hunk 7]
```diff
@@ -228,6 +239,11 @@ func (g *ConfigGenerator) Daemonset() *appsv1.DaemonSet {
 						VolumeSource: corev1.VolumeSource{
 							EmptyDir: &corev1.EmptyDirVolumeSource{},
 						},
+					}, {
+						Name: authenticatorVolumeName,
+						VolumeSource: corev1.VolumeSource{DownwardAPI: &corev1.DownwardAPIVolumeSource{
+							Items: []corev1.DownwardAPIVolumeFile{{Path: "labels",
+								FieldRef: &corev1.ObjectFieldSelector{FieldPath: "metadata.labels"}}}}},
```
Comment (lack) [inline pkg/configgen/configgen.go:247]: This is cool!  What are we using these labels for?

File: pkg/hostap/monitor.go (hunk 6 above)
Comment (pperiyasamy) [reply pkg/hostap/monitor.go:187]: done.

File: pkg/configgen/configgen.go (hunk 7 above)
Comment (pperiyasamy) [reply pkg/configgen/configgen.go:247]: This is to inject authenticator's crd name inside authenticator pod and then used inside later for updating status and recording events on it.

File: monitor/main.go (hunk 4 above)
Comment (pperiyasamy) [reply monitor/main.go]: yes, done it now.

File: api/v1/authenticator_types.go (hunk 1 above)
Comment (pperiyasamy) [reply api/v1/authenticator_types.go]: moved it under controller code now, hope that's ok deploy rbac configuration from the controller.

File: internal/k8s/k8s.go
Comment (pperiyasamy) [reply internal/k8s/k8s.go:1]: you could find some references w.r.t logic, but this is not copied from some place as exactly.

File: api/v1/authenticator_types.go (hunk 3 above)
Comment (pperiyasamy) [reply api/v1/authenticator_types.go]: yes, done.

File: pkg/hostap/monitor.go (hunk 5 above)
Comment (pperiyasamy) [reply pkg/hostap/monitor.go]: I was just using event string (which contains 'state) as it is to denotate state in k8s. anyway i changed it to use enum now.

File: api/v1/authenticator_types.go [hunk 8]
```diff
@@ -20,6 +20,19 @@ import (
...
+
+var (
+	IfStateUninitialized IfState = "Uninitialized"
+	IfStateDisabled      IfState = "Disabled"
+	IfStateCountryUpdate IfState = "CountryUpdate"
+	IfStateAcs           IfState = "ACS"
+	IfStateHtScan        IfState = "HT Scan"
+	IfStateDfs           IfState = "DFS"
+	IfStateEnabled       IfState = "Enabled"
+	IfStateUnknown       IfState = "Unknown"
```
Comment (lack) [inline api/v1/authenticator_types.go:33]: Are all of these valid for 802.1x, or are some only valid for WiFi authentication (which is the other thing hostapd does)?  Can we get a one-liner about what each one means?

File: api/v1/authenticator_types.go (hunk 1 above)
Comment (lack) [reply api/v1/authenticator_types.go]: I think we can start that way, and expose this later if needed.

File: api/v1/authenticator_types.go (hunk 8 above)
Comment (pperiyasamy) [reply api/v1/authenticator_types.go:33]: sure, will do it with another commit.

**Pull request merged**